File: mobile/composeApp/src/commonMain/kotlin/com/tamixa/ui/StoryListenerUi.kt
package com.tamixa.ui

import com.tamixa.domain.Story
import com.tamixa.util.TamixaConstants

private val funStoryLabels = setOf("fun stories", "funny stories")

/** Lighter / classic lane — match admin categories "Fun stories" or "Funny Stories". */
fun isFunStory(theme: String, category: String?): Boolean {
    val c = category?.trim()?.lowercase().orEmpty()
    val t = theme.trim().lowercase()
    return c in funStoryLabels || t in funStoryLabels
}

fun isFunStory(story: Story): Boolean = isFunStory(story.theme, story.category)

/** Legacy "Learn ·" hub rows; prefer [isFunStory] for badges. */
private val learnPrefix = Regex("^learn\\b", RegexOption.IGNORE_CASE)

fun isLearnStory(theme: String, category: String?): Boolean {
    val c = category?.trim().orEmpty()
    val t = theme.trim()
    return learnPrefix.containsMatchIn(c) || learnPrefix.containsMatchIn(t)
}

fun isLearnStory(story: Story): Boolean = isLearnStory(story.theme, story.category)

/** Edu & safety lane: Learn-prefixed metadata or explicit Digital Safety series. */
fun isLearnOrDigitalSafetyStory(story: Story): Boolean {
    if (isLearnStory(story)) return true
    val needle = "digital safety"
    return story.theme.contains(needle, ignoreCase = true) ||
        story.category?.contains(needle, ignoreCase = true) == true
}

/** Interactive life-simulator series (branching); theme/category starts with "Learn · Simulator". */
private val simulatorPrefix = Regex("^learn\\s*·\\s*simulator\\b", RegexOption.IGNORE_CASE)

fun isSimulatorStory(theme: String, category: String?): Boolean {
    val c = category?.trim().orEmpty()
    val t = theme.trim()
    return simulatorPrefix.containsMatchIn(c) || simulatorPrefix.containsMatchIn(t)
}

fun isSimulatorStory(story: Story): Boolean = isSimulatorStory(story.theme, story.category)

/** Library “Practice” hub: naming convention or non-empty interactive graph from API. */
fun isInteractivePracticeLibraryStory(story: Story): Boolean {
    if (isSimulatorStory(story)) return true
    val g = story.interactiveGraph
    return g != null && g.isNotEmpty()
}

private const val MAX_CONTEXT = 120

private fun truncate(s: String): String =
    if (s.length > MAX_CONTEXT) s.take(MAX_CONTEXT) + "…" else s

/** Optional subtitle under the player title while listening. */
fun listenerPlaybackSubtitle(story: Story, storySource: String): String? {
    val theme = story.theme.trim()
    val category = story.category?.trim().orEmpty()
    val title = story.title?.trim().orEmpty()
    return when (storySource) {
        TamixaConstants.STORY_SOURCE_LIBRARY -> {
            if (isFunStory(story)) {
                val focus = category.takeIf { it.isNotBlank() } ?: theme.takeIf { it.isNotBlank() } ?: "Just for fun"
                truncate(focus)
            } else if (theme.isNotBlank() && title.isNotBlank() && theme != title) {
                truncate(theme)
            } else if (category.isNotBlank()) {
                truncate(category)
            } else {
                null
            }
        }
        else -> {
            if (theme.isBlank()) null
            else if (title.isNotBlank() && theme == title) null
            else truncate(theme)
        }
    }
}

File: web/src/lib/storyListenerUi.ts
/**
 * Listener-facing copy: every Tamixa library story is built for growth (heart, vocabulary, curiosity).
 * The "Fun" lane is for lighter / classic tales—use category "Fun stories" or "Funny Stories" in admin.
 */

const FUN_STORY_LABELS = new Set(["fun stories", "funny stories"]);

/** True when theme or category is the curated fun / laughs lane (legacy-friendly). */
export function isFunStory(theme: string, category?: string | null): boolean {
  const c = category?.trim().toLowerCase() ?? "";
  const t = theme?.trim().toLowerCase() ?? "";
  return FUN_STORY_LABELS.has(c) || FUN_STORY_LABELS.has(t);
}

/** @deprecated Prefer {@link isFunStory}; kept for any code still checking old "Learn ·" rows. */
const LEARN_PREFIX = /^learn\b/i;

export function isLearnStory(theme: string, category?: string | null): boolean {
  const c = category?.trim() ?? "";
  const t = theme?.trim() ?? "";
  return LEARN_PREFIX.test(c) || LEARN_PREFIX.test(t);
}

const DIGITAL_SAFETY = "digital safety";

/** Learn-prefixed rows or explicit Digital Safety series (library Edu lane). */
export function isLearnOrDigitalSafetyStory(theme: string, category?: string | null): boolean {
  if (isLearnStory(theme, category)) return true;
  const t = theme?.trim().toLowerCase() ?? "";
  const c = category?.trim().toLowerCase() ?? "";
  return t.includes(DIGITAL_SAFETY) || c.includes(DIGITAL_SAFETY);
}

const SIMULATOR_PREFIX = /^learn\s*·\s*simulator\b/i;

/** Theme/category convention for Learn · Simulator series. */
export function isSimulatorStory(theme: string, category?: string | null): boolean {
  const c = category?.trim() ?? "";
  const t = theme?.trim() ?? "";
  return SIMULATOR_PREFIX.test(c) || SIMULATOR_PREFIX.test(t);
}

/**
 * Practice hub: simulator naming or non-empty interactive graph payload from API.
 *
 * Canonical graph for parents: `GET /v1/stories/library/{id}` returns a merged `interactiveGraph` and does not
 * expose `translationInteractiveGraphOverlay` (admin-only overlay). Use `interactiveGraph` for branching UI.
 */
export function isInteractivePracticeLibraryStory(story: {
  theme: string;
  category?: string | null;
  interactiveGraph?: unknown;
}): boolean {
  if (isSimulatorStory(story.theme, story.category)) return true;
  const g = story.interactiveGraph;
  if (g == null) return false;
  if (typeof g === "object" && !Array.isArray(g) && Object.keys(g as object).length > 0) return true;
  return false;
}

export function funCornerBadgeLabel(): string {
  return "Just for fun";
}

/** Library poster chip for branching Edu / simulator tales (matches mobile copy). */
export function interactivePracticeBadgeLabel(): string {
  return "You choose";
}

/** Short line under card title. */
export function listenerCardMetaLine(theme: string, category?: string | null): string {
  const c = category?.trim();
  if (c) return c;
  return theme.trim();
}

const MAX_PLAYER_CONTEXT = 120;

/** Subtitle under the player title while listening. */
export function listenerPlaybackSubtitle(
  storySource: string,
  theme: string,
  category?: string | null,
  title?: string | null,
): string | null {
  const t = theme?.trim() ?? "";
  const c = category?.trim() ?? "";
  const ti = title?.trim() ?? "";
  if (storySource === "library" || storySource === "favorites") {
    if (isFunStory(t, c || null)) {
      const focus = c || t || "Just for fun";
      return focus.length > MAX_PLAYER_CONTEXT ? `${focus.slice(0, MAX_PLAYER_CONTEXT)}…` : focus;
    }
    if (t && ti && t !== ti) {
      return t.length > MAX_PLAYER_CONTEXT ? `${t.slice(0, MAX_PLAYER_CONTEXT)}…` : t;
    }
    if (c) {
      return c.length > MAX_PLAYER_CONTEXT ? `${c.slice(0, MAX_PLAYER_CONTEXT)}…` : c;
    }
    return null;
  }
  if (storySource === "generated" || storySource === "mine") {
    if (!t) return null;
    if (ti && t === ti) return null;
    return t.length > MAX_PLAYER_CONTEXT ? `${t.slice(0, MAX_PLAYER_CONTEXT)}…` : t;
  }
  return null;
}
